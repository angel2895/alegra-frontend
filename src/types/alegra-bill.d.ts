export interface AlegraBill {
  date: Date;
  dueDate: Date;
  client: number;
  seller: number;
  items: Item[];
}

export interface Item {
  id: number;
  description: string;
  price: number;
  quantity: number;
}
