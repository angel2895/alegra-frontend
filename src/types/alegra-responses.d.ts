export interface AlegraResponseProduct {
  id: string;
  category: Category;
  name: string;
  description: string;
  reference: string;
  status: string;
  price: Price[];
  inventory: Inventory;
  tax: any[];
  customFields: any[];
  productKey: null;
  type: string;
  itemType: null;
  settingsOnPos: SettingsOnPos;
}

export interface Category {
  id: string;
  name: string;
}

export interface Inventory {
  unit: string;
}

export interface Price {
  idPriceList: number;
  name: string;
  type: string;
  price: string;
  currency: Currency;
}

export interface Currency {
  code: string;
  symbol: string;
}

export interface SettingsOnPos {
  hide: null;
}

export interface AlegraResponseBill {
  id: string;
  date: Date;
  dueDate: Date;
  datetime: Date;
  observations: null;
  anotation: null;
  termsConditions: string;
  status: string;
  client: Client;
  numberTemplate: NumberTemplate;
  subtotal: number;
  discount: number;
  tax: number;
  total: number;
  totalPaid: number;
  balance: number;
  decimalPrecision: string;
  warehouse: Warehouse;
  operationType: string;
  seller: Seller;
  priceList: PriceList;
  items: Item[];
  costCenter: null;
}

export interface Client {
  id: string;
  name: string;
  identification: string;
  phonePrimary: null;
  phoneSecondary: null;
  fax: null;
  mobile: string;
  email: string;
  address: Address;
}

export interface Address {
  state: string;
  city: string;
  district: string;
  urbanization: null;
  ubigeoCode: string;
  address: string;
}

export interface Item {
  name: string;
  description: null;
  price: number;
  discount: number;
  reference: null;
  quantity: number;
  id: number;
  productKey: null;
  unit: Unit;
  tax: any[];
  total: number;
}

export enum Unit {
  Unit = "unit",
}

export interface NumberTemplate {
  id: string;
  prefix: string;
  number: string;
  text: null;
  documentType: string;
  fullNumber: string;
  formattedNumber: string;
}

export interface PriceList {
  id: number;
  name: string;
}

export interface Seller {
  id: number;
  name: string;
  identification: string;
  observations: string;
}

export interface Warehouse {
  id: string;
  name: string;
}
