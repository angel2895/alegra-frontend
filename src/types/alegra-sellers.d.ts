export interface AlegraSeller {
  id: number;
  name: string;
  identification: string;
  observations: string;
  status: Status;
}

export enum Status {
  Active = "active",
}
