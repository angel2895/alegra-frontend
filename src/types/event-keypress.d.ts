export interface EventKeyPress {
  isTrusted: boolean;
  altKey: boolean;
  bubbles: boolean;
  cancelBubble: boolean;
  cancelable: boolean;
  charCode: Number;
  code: String;
}
