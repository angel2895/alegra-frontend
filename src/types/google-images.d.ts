export interface GoogleImages {
  search_metadata: SearchMetadata;
  search_parameters: SearchParameters;
  search_information: SearchInformation;
  suggested_searches: SuggestedSearch[];
  images_results: GoogleImage[];
}

export interface GoogleImage {
  position: number;
  thumbnail: string;
  source: string;
  title: string;
  link: string;
  original: string;
  is_product: boolean;
  in_stock?: boolean;
}

export interface SearchInformation {
  image_results_state: string;
  query_displayed: string;
}

export interface SearchMetadata {
  id: string;
  status: string;
  json_endpoint: string;
  created_at: string;
  processed_at: string;
  google_url: string;
  raw_html_file: string;
  total_time_taken: number;
}

export interface SearchParameters {
  engine: string;
  q: string;
  google_domain: string;
  ijn: string;
  device: string;
  tbm: string;
}

export interface SuggestedSearch {
  name: string;
  link: string;
  chips: string;
  serpapi_link: string;
  thumbnail: string;
}
