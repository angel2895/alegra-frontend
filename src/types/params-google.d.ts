export interface ParamsGoogle {
  api_key: string;
  tbm: string;
  ijn: number;
  q?: string;
}
