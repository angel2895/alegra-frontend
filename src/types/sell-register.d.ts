export interface SellRegister {
  id: number;
  name: string;
  points: number;
  products: Image[];
}

export interface SellRegisterShort {
  id: number;
  name: string;
  points: number;
}

export interface Image {
  key: string;
  url: string;
  title: string;
}

export interface ImageSelled {
  id: number;
  name: string;
  key: string;
  url: string;
  title: string;
}
