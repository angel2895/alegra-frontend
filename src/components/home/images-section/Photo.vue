<template>
  <div class="container-image">
    <img
      src="@/assets/default/default.jpg"
      alt="image-default"
      v-if="getLoading"
    />
    <img
      v-else
      :src="this.image.thumbnail"
      alt="image-search"
      :class="{ loadingImage: image }"
      @click="selectImage"
    />

    <div
      class="container-author"
      :class="{ loadingAuthor: image }"
      v-if="image && !getLoading"
    >
      <span class="author-name">{{ seller.name }}</span>
    </div>
    <div class="container-buy" v-if="image && !getLoading">
      <img
        v-if="!imageSelected"
        src="@/assets/icons/heart-void.svg"
        alt=""
        class="heart-void"
        @click="selectImage"
      />
      <img
        v-else
        src="@/assets/icons/heart-fill.svg"
        alt=""
        class="heart-fill"
        @click="selectImage"
      />
    </div>
  </div>
</template>

<script lang="ts">
// Libraries
import { Options, Vue } from "vue-class-component";
import { mapGetters } from "vuex";
import { mapMutations } from "vuex";

// Types
import { GoogleImage } from "@/types/google-images";
import { AlegraSeller } from "@/types/alegra-seller";
import { ImageSelled } from "@/types/sell-register";

@Options({
  props: {
    image: {} as GoogleImage,
    seller: {} as AlegraSeller,
  },
  watch: {
    getLoading(loading) {
      if (!loading) {
        // reset the heart
        this.imageSelected = false;
      }
    },
  },
  computed: mapGetters(["getLoading"]),
  methods: mapMutations(["UPDATE_LIST_SELL_REGISTER"]),
})
export default class Photo extends Vue {
  image!: GoogleImage;
  seller!: AlegraSeller;

  imageLoaded = false;
  imageSelected = false;

  UPDATE_LIST_SELL_REGISTER!: (payload: ImageSelled) => void;

  selectImage(): void {
    this.imageSelected = !this.imageSelected;

    const imageSelled: ImageSelled = {
      id: 0,
      name: "",
      key: "",
      url: "",
      title: "",
    };

    imageSelled.id = this.seller.id;
    imageSelled.name = this.seller.name;
    imageSelled.key = this.image.original;
    imageSelled.url = this.image.original;
    imageSelled.title = this.image.title;

    this.UPDATE_LIST_SELL_REGISTER(imageSelled);
  }
}
</script>

<style scoped>
.container-image {
  display: block;
  padding: 0px;
  height: 140px;
}
.container-image > img {
  border: 1px solid rgba(100, 100, 100, 0.3);
  border-radius: 5px;
  height: 120px;
}
.loadingImage {
  animation: fadein 2s;
}
@keyframes fadein {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}
.container-author {
  position: relative;
  top: -40px;

  z-index: 2;
}
.author-name {
  padding: 5px 10px;
  margin-left: 20px;
  width: 10px;

  font-size: 1.2rem;
  font-weight: 500;

  color: white;
  background-color: rgba(44, 44, 44, 0.7);
}
.loadingAuthor {
  animation: slide 2s;
}
@keyframes slide {
  0% {
    transform: translateX(-20px);
  }
  100% {
    transform: translateX(0px);
  }
}
.container-buy {
  position: relative;
  top: -150px;
  padding-right: 20px;

  text-align: end;

  z-index: 2;
}
.heart-void,
.heart-fill {
  background: radial-gradient(
    rgba(0, 0, 0, 0.6),
    rgba(0, 0, 0, 0.5),
    rgba(0, 0, 0, 0.3),
    rgba(0, 0, 0, 0.1),
    rgba(0, 0, 0, 0),
    rgba(0, 0, 0, 0),
    rgba(0, 0, 0, 0)
  );

  animation: fill-heart 1s;
}
@keyframes fill-heart {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}

@media (min-width: 768px) {
  .container-image {
    height: 270px;
  }
  .container-image > img {
    height: 250px;
  }
  .container-buy {
    top: -250px;
  }
}

@media (min-width: 1300px) {
  .container-image {
    height: 320px;
  }
  .container-image > img {
    height: 300px;
  }
  .container-buy {
    top: -300px;
  }
}
</style>
