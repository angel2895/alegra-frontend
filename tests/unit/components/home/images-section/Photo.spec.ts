import { shallowMount } from "@vue/test-utils";
import { createStore } from "vuex";
// Components
import Photo from "@/components/home/images-section/Photo.vue";
// Types
import { GoogleImage } from "@/types/google-images";
import { AlegraSeller } from "@/types/alegra-seller";

const store = createStore({
  state() {
    return { isLoading: true };
  },
  getters: {
    getLoading: (state) => (state as any).isLoading,
  },
  mutations: {
    // this fake function is only nessesary because in the current version setData
    // function is failing in typescript, jest and vue3, in near future will be solved this fix
    UPDATE_LIST_SELL_REGISTER(state) {
      (state as any).isLoading = false;
    },
  },
});

const wrapper = shallowMount(Photo, {
  global: {
    plugins: [store],
  },
  props: {
    image: {
      position: 1,
      thumbnail:
        "https://serpapi.com/searches/62168bbe5c10595dfa382d6a/images/05109f1d08e78809ac8ac33e43ac30be36fe799ed67b01bb8b42b1203f73ed6e.png",
      source: "en.wikipedia.org",
      title: "History of Apple Inc. - Wikipedia",
      link: "https://en.wikipedia.org/wiki/History_of_Apple_Inc.",
      original:
        "https://upload.wikimedia.org/wikipedia/commons/f/fa/Apple_logo_black.svg",
      is_product: false,
    } as GoogleImage,
    seller: {
      id: 11,
      name: "Alexander Benavides",
      identification: "1c25ad194ba17a2bc5c07c62c7d66350",
      observations:
        'Alexander Benavides es un/una gran vendedor/a de nuestra empresa "Imagenes del Mundo".',
      status: "active",
    } as AlegraSeller,
  },
});

describe("Photo.vue", () => {
  it("Match with snapshot", (): void => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("Until charge the image only show the default image", (): void => {
    const $imageDefault = wrapper.find("img");
    expect($imageDefault.attributes("alt")).toBe("image-default");
    expect($imageDefault.attributes("src")).toBe(undefined);
  });

  it("Contains (image, author, heart) after loading with their atributes", async (): Promise<void> => {
    await wrapper.vm.UPDATE_LIST_SELL_REGISTER();
    const [$image, $heart] = wrapper.findAll("img");
    const $authorContainer = wrapper.find(".container-author");
    const $author = wrapper.find(".author-name");
    // Main iamge
    expect($image.exists()).toBeTruthy();
    expect($image.attributes("alt")).toBe("image-search");
    expect($image.attributes("class")).toBe("loadingImage");
    expect($image.attributes("src")).toBe(wrapper.vm.image.thumbnail);
    // Heart void
    expect($heart.exists()).toBeTruthy();
    expect($heart.attributes("class")).toBe("heart-void");
    expect($heart.attributes("src")).toBe(undefined);
    // Author Container
    expect($authorContainer.exists()).toBeTruthy();
    expect($authorContainer.attributes("class")).toBe(
      "container-author loadingAuthor"
    );
    // Author
    expect($author.exists()).toBeTruthy();
    expect($author.text()).toBe(wrapper.vm.seller.name);
  });

  it("Fill heart when the image is clicked", async (): Promise<void> => {
    await wrapper.vm.UPDATE_LIST_SELL_REGISTER();

    const $image = wrapper.findAll("img")[0];

    await $image.trigger("click");

    const $heartFilled = wrapper.findAll("img")[1];

    expect($heartFilled.exists()).toBeTruthy();
    expect($heartFilled.attributes("class")).toBe("heart-fill");
  });

  it("Unfill the heart when the image is clicked again", async (): Promise<void> => {
    await wrapper.vm.UPDATE_LIST_SELL_REGISTER();

    const $image = wrapper.findAll("img")[0];

    // Because the mount is global the last click is conserved so its only nessesary 1 more click
    await $image.trigger("click");

    const $heart = wrapper.findAll("img")[1];

    expect($heart.exists()).toBeTruthy();
    expect($heart.attributes("class")).toBe("heart-void");
  });
});
