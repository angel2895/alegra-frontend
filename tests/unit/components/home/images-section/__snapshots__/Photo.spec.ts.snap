// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photo.vue Match with snapshot 1`] = `
<div class="container-image"><img alt="image-default">
  <!--v-if-->
  <!--v-if-->
</div>
`;
